sw/device/tests/sim_dv/uart_tx_rx_test.c

```diff
@@ -0,0 +1,350 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/uart.h"
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/log.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_plic.h"
+#include "sw/device/lib/dif/dif_uart.h"
+#include "sw/device/lib/handler.h"
+#include "sw/device/lib/irq.h"
+#include "sw/device/lib/runtime/check.h"
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/testing/test_status.h"
+
+#define UART_DATASET_SIZE 128
+#define UART0_BASE_ADDR 0x40000000u
+#define PLIC0_BASE_ADDR 0x40090000u
+
+static dif_uart_t uart0;
+static dif_plic_t plic0;
+
+/**
+ * UART TX RX test
+ *
+ * This test sends and receives a known dataset over UART. The size of the
+ * dataset is indicated with UART_DATASET_SIZE. The dataset is agreed upon by
+ * the device (a.k.a. the OpenTitan chip) and the host (a.k.a. the simulation
+ * device, such as DV testbench). communicating with it. Data transmitted over
+ * TX is checked for correctness at the host, and likewise, data sent by the
+ * host is checked for correctness at the device (in this SW test). The data
+ * transmitted over TX and RX ports may occur simultaneously. The test ensures
+ * that the TX watermark, RX watermark and TX empty interrupts are seen.
+ * At the end, the host transmits another set of random data (greater than the
+ * RX fifo size which the device drops, to generate the RX overflow condition.
+ * The test passes when the datasets at both ends match the expected and all
+ * required intererupts are seen.
+ */
+
+// TODO: Make these datasets random.
```

Comment:
Could you file an issue for this? I'll need to set up some kind of "inject random seed" mechanism in `device.h` for this to work.

Reply:
> Could you file an issue for this? I'll need to set up some kind of "inject random seed" mechanism in `device.h` for this to work.

That's part of it, but the communication needed by both the SW and the DV side is hard to coordinate. Methods I've seen include random numbers blasted into a C- and SV-readable file; or C-containers read by .c files and through DPIs; but never seen an elegant way to do this.